{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "''' Manipulate Multi-Sample VCF files in Pandas (Python3). \n",
    "'''\n",
    "from collections import OrderedDict\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import sys\n",
    "import os\n",
    "import collections\n",
    "\n",
    "# path to this %%file\n",
    "if sys.platform == \"win32\":\n",
    "    file_path = os.path.dirname(os.path.abspath(\"__file__\"))+\"\\\\\"\n",
    "else:\n",
    "    file_path = os.path.dirname(os.path.abspath(\"__file__\"))+\"/\"\n",
    "\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.max_colwidth', -1)\n",
    "\n",
    "\n",
    "def vcf2dataframe(filename, genotype_level=True, info_level=True, UID=False):\n",
    "    '''Open a VCF file and returns a MultiIndex pandas.DataFrame.\n",
    "    Args:\n",
    "        filename: vcf file to be converted to a dataframe\n",
    "        genotype_level: place the genotype information into a second level column index\n",
    "        info_level: place the info IDs into a second level column index\n",
    "        UID: rename index to a unique variant identifier\n",
    "    Notes:\n",
    "        having any of these variables set to True will result\n",
    "        in the DataFrame being generated very slowly. This is\n",
    "        especially true for the UID variable.\n",
    "    '''\n",
    "    if filename.endswith(\".gz\"):\n",
    "        raise IOError(\"pdVCF does not support compressed VCF files.\")\n",
    "\n",
    "    # get INFO fields and Headers as lists\n",
    "    VCF_HEADER = get_vcf_header(filename)\n",
    "    INFO_FIELDS = get_info_fields(filename)\n",
    "    print(INFO_FIELDS)\n",
    "\n",
    "    # Count how many comment lines should be skipped.\n",
    "    comments = count_comments(filename)\n",
    "\n",
    "    # Return a simple dataframe representative of the VCF data.\n",
    "    df = pd.read_table(filename, skiprows=comments,\n",
    "                       names=VCF_HEADER, usecols=range(len(VCF_HEADER)))\n",
    "\n",
    "    if genotype_level:\n",
    "        df = get_genotype_data(df)\n",
    "\n",
    "    if info_level:\n",
    "        df = get_info_data(df, INFO_FIELDS)\n",
    "\n",
    "    if UID:\n",
    "        df = index2UID(df)\n",
    "    \n",
    "    # replace empty cells\n",
    "    df = df.replace(\"\", np.nan)\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def get_vcf_header(filename):\n",
    "    ''' Get all header names from a given VCF file and return as a list.\n",
    "    '''\n",
    "    with open(filename) as input_file:\n",
    "        row = [x for x in input_file if not x.startswith('##')] # skip unwanted headers\n",
    "        head = next(iter(row))    # generator to deal with the header line only.\n",
    "        split_head = [re.sub(r'#|\\n', '', x) for x in head.split(\"\\t\")]\n",
    "        return split_head\n",
    "\n",
    "\n",
    "def get_info_fields(filename):\n",
    "    ''' Get all ID names in the given VCFs INFO field and return as a list.\n",
    "    '''\n",
    "    with open(filename) as input_file:\n",
    "        row = [x for x in input_file if x.startswith('##INFO')]\n",
    "        info_fields = [x[11:].split(',')[0] for x in row]\n",
    "        return info_fields\n",
    "\n",
    "\n",
    "\n",
    "def count_comments(filename):\n",
    "    ''' Count all lines in a given VCF file starting with #.\n",
    "    '''\n",
    "    comments = 0\n",
    "    with open(filename) as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#'):\n",
    "                comments += 1\n",
    "\n",
    "            else:\n",
    "                break\n",
    "\n",
    "    return comments\n",
    "\n",
    "\n",
    "\n",
    "def replace_series_strings(df, col, dic, substring):\n",
    "    ''' Replace the the keys with the items of the given\n",
    "        dictionary for all strings or substrings in a\n",
    "        given column\n",
    "    Args:\n",
    "        col: column name to replace strings\n",
    "        dic: dictionary where the key is the string to replace with the item\n",
    "        substrings: search and replace for either substrings (True) or exact strings (False)\n",
    "    Returns:\n",
    "        dataframe with the given column having all the\n",
    "        entries identified as the key in the given dict\n",
    "        replaced with the item in said dict\n",
    "    '''\n",
    "    if not isinstance(substring, bool):\n",
    "        raise TypeError(\"substring argument must equal True or False\")\n",
    "\n",
    "    for string, correction in dic.items():\n",
    "        if substring is True:\n",
    "            df[col] = df[col].str.replace(string, correction)\n",
    "        elif substring is False:\n",
    "            df[col] = df[col].replace(string, correction, regex=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "def get_genotype_data(df):\n",
    "    ''' Give each sample column a second level column for every field\n",
    "        detailed in the FORMAT column and return as a MultIndex dataframe.\n",
    "    Args:\n",
    "        df: DataFrame deriving from a VCF via vcf2dataframe()\n",
    "    '''\n",
    "    # contain the variant columns and the sample names in seperate lists\n",
    "    normal = list(df.iloc[:, :9].columns)\n",
    "    samples = list(df.iloc[:, 9:].columns)\n",
    "    form = df['FORMAT'].str.split(\":\")[0]\n",
    "\n",
    "    # These columns remain the same\n",
    "    remain = pd.DataFrame(data=df[normal].values,\n",
    "                          columns=pd.MultiIndex.from_tuples(\n",
    "                            [(x, '') for x in normal] ))\n",
    "\n",
    "    # list of dataframes where every sample has sub columns for each genotype info\n",
    "    sams = [pd.DataFrame(data=list(df[col].str.split(':').dropna()),\n",
    "                         columns=pd.MultiIndex.from_product([ [col], form ]))\n",
    "            for col in samples]\n",
    "    \n",
    "    # add allele balance to sample genotype information\n",
    "    sams = [calc_AB(sam) for sam in sams]\n",
    "    \n",
    "    # concat all dfs in the list\n",
    "    df2 = pd.concat([remain] + sams, axis=1)\n",
    "\n",
    "    return df2\n",
    "\n",
    "\n",
    "\n",
    "def calc_AB(vcf):\n",
    "    ''' Calculate allele balance for all samples in a given \n",
    "        pdVCF. Also converts DP & GQ to numeric type.\n",
    "    \n",
    "    Args:\n",
    "        vcf: pdVCF with genotype information extracted\n",
    "        \n",
    "    Notes:\n",
    "        ONLY WORKS FOR BIALLELIC VARIANTS\n",
    "    '''\n",
    "    sam = vcf.columns.levels[0][0]\n",
    "    vcf[sam,'DP'] = pd.to_numeric(vcf[sam,'DP'])\n",
    "    vcf[sam,'GQ'] = pd.to_numeric(vcf[sam,'GQ'])\n",
    "    AD = vcf.xs('AD', level=1, axis=1).unstack().str.split(\",\", n=2)\n",
    "    DP = vcf.xs('DP', level=1, axis=1).unstack()\n",
    "    AB = round(pd.to_numeric(AD.str[1]) / pd.to_numeric(DP), 2)\n",
    "    vcf[sam, 'AB'] = AB.tolist()\n",
    "        \n",
    "    return vcf\n",
    "\n",
    "\n",
    "\n",
    "def get_info_data(df, info_fields):\n",
    "    ''' Transform the INFO IDs into second level column indexes and return\n",
    "        the df as a MultiIndex dataframe.\n",
    "    Args:\n",
    "        df: DataFrame deriving from a VCF via vcf2dataframe()\n",
    "        info_fields: a list of all the INFO IDs in the given df\n",
    "    '''\n",
    "    # Alter Info field for some variables that don't work well\n",
    "    df['INFO'] = df['INFO'].str.replace(\";DB\",\";DB=1\")\n",
    "    df['INFO'] = df['INFO'].str.replace(\";STR\",\";STR=1\")\n",
    "\n",
    "    # identify Info fields not present in each row and fill them with a 0\n",
    "    for name in info_fields:\n",
    "        if name == info_fields[0]:\n",
    "            name = \"{}=\".format(name)\n",
    "        else:\n",
    "            name = \";{}=\".format(name)\n",
    "\n",
    "        not_present = df['INFO'][~df.INFO.str.contains(name)].add(\"{}0\".format(name))\n",
    "        present = df['INFO'][df.INFO.str.contains(name)]\n",
    "        df['INFO'] = not_present.append(present).sort_index()\n",
    "\n",
    "    # reorder INFO fields so they are are all in the same order\n",
    "    df['INFO'] = df['INFO'].apply(lambda x: ';'.join(elem for elem in sorted(x.split(\";\"))))\n",
    "\n",
    "    # remove all info_field names from the info values, starting with the info field with the longest name first\n",
    "    unwanted = info_fields + ['=']\n",
    "    unwanted.sort(key=len, reverse=True)\n",
    "    remove = collections.OrderedDict([(x, '') for x in unwanted])\n",
    "    df = replace_series_strings(df, col='INFO', dic=remove, substring=True)\n",
    "\n",
    "    # create a new multi-index df containing only the info fields with the IDs as the second level\n",
    "    info = pd.DataFrame(data=list(df['INFO'].str.split(';')),\n",
    "                        columns=pd.MultiIndex.from_product([ ['INFO'], info_fields]))\n",
    "\n",
    "    if not isinstance(df.columns, pd.MultiIndex):\n",
    "        # create another multi-index df without the info fields where the second level is nothing\n",
    "        df = pd.DataFrame(data=df.drop('INFO', axis=1).values,\n",
    "                              columns=pd.MultiIndex.from_tuples(\n",
    "                                [(x, '') for x in list(df.drop('INFO', axis=1).columns)] ))\n",
    "\n",
    "    else:\n",
    "        df = df.drop('INFO', axis=1)\n",
    "\n",
    "    variant = df.iloc[:, :8]\n",
    "    samples = df.iloc[:, 8:]\n",
    "\n",
    "    # replace the info fields in the original df with the multi-index df created above\n",
    "    final_df = pd.concat([variant] + [info] + [samples], axis=1)\n",
    "\n",
    "    # MQ and MQ0 are in the wrong order so name swapping is required\n",
    "    if 'MQ0' in info_fields and 'MQ' in info_fields:\n",
    "        final_df = final_df.rename(columns={'MQ0': 'TEMP', 'MQ': 'MQ0'})\n",
    "        final_df = final_df.rename(columns={'TEMP': 'MQ'})\n",
    "\n",
    "    return final_df\n",
    "\n",
    "\n",
    "\n",
    "def index2UID(df):\n",
    "    ''' Replace the index with a unique variant identifier.\n",
    "    '''\n",
    "    if isinstance(df.columns, pd.MultiIndex):\n",
    "        UID = df.apply(lambda x: \"{}:{}-{}/{}\".format(x['CHROM'][0], x['POS'][0],\n",
    "                                                      x['REF'][0], x['ALT'][0]), axis=1)\n",
    "    else:\n",
    "         UID = df.apply(lambda x: \"{}:{}-{}/{}\".format(x['CHROM'], x['POS'],\n",
    "                                                  x['REF'], x['ALT']), axis=1)\n",
    "\n",
    "    df['UID'] = UID\n",
    "\n",
    "    if df['UID'].value_counts()[0] > 1:\n",
    "        raise ValueError(\"The UID is not unique.\")\n",
    "        \n",
    "    # remove UID column (UID now only accessible via index)  \n",
    "    df = df.drop('UID', axis=1)\n",
    "\n",
    "    return df.rename(UID)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "j = \"var.both.taadUkJan2017.filters.vcf\"\n",
    "o = \"var.both.taadUkOctoberRepeatLibrariesDec2016.filters.vcf\"\n",
    "t = \"26PL1207.oct.noInDels.recode.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.4514596462249756\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start = time.time()\n",
    "\n",
    "vcf2dataframe(j, genotype_level=False, \n",
    "             info_level=False, UID=False)\n",
    "\n",
    "done = time.time() - start\n",
    "print(done)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AC', 'AF', 'AN', 'BaseQRankSum', 'DB', 'DP', 'DS', 'Dels', 'ExcessHet', 'FS', 'HaplotypeScore', 'InbreedingCoeff', 'MLEAC', 'MLEAF', 'MQ', 'MQ0', 'MQRankSum', 'QD', 'RPA', 'RU', 'ReadPosRankSum', 'SOR', 'STR', 'set']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:220: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:253: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1:2160444-A/G    0.341\n",
       "1:2160448-G/A    0.032\n",
       "1:2160449-C/T    0.010\n",
       "1:2160451-G/A    0.021\n",
       "1:2160455-C/T    0.010\n",
       "Name: AF, dtype: object"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf = vcf2dataframe(o, genotype_level=True, \n",
    "             info_level=True, UID=True)\n",
    "\n",
    "#vcf[(vcf['21AI1224']['AB'] > 0.3) & (vcf['21AI1224']['DP'] > 49) & (vcf['21AI1224']['GQ'] > 29)]['21AI1224'].head()\n",
    "vcf['INFO'].apply(pd.to_numeric, errors='ignore')['AF'].head()\n",
    "\n",
    "#vcf['INFO']['AF'] > 12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:219: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:252: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th colspan=\"2\" halign=\"left\">INFO</th>\n",
       "      <th colspan=\"6\" halign=\"left\">26PL1207</th>\n",
       "      <th colspan=\"6\" halign=\"left\">22MI1099</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AC</th>\n",
       "      <th>LOLZ</th>\n",
       "      <th>GT</th>\n",
       "      <th>AD</th>\n",
       "      <th>DP</th>\n",
       "      <th>GQ</th>\n",
       "      <th>PL</th>\n",
       "      <th>AB</th>\n",
       "      <th>GT</th>\n",
       "      <th>AD</th>\n",
       "      <th>DP</th>\n",
       "      <th>GQ</th>\n",
       "      <th>PL</th>\n",
       "      <th>AB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1:1000-A/G</th>\n",
       "      <td>1</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.89</td>\n",
       "      <td>LowQual</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,1</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>3,0,197</td>\n",
       "      <td>0.09</td>\n",
       "      <td>./.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2234385-C/T</th>\n",
       "      <td>1</td>\n",
       "      <td>2234385</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9953.67</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>0/1</td>\n",
       "      <td>5,11</td>\n",
       "      <td>16</td>\n",
       "      <td>99</td>\n",
       "      <td>289,0,103</td>\n",
       "      <td>0.69</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,1</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3,0,197</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2235243-C/T</th>\n",
       "      <td>1</td>\n",
       "      <td>2235243</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>82.06</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>4,1</td>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "      <td>16,0,102</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0/1</td>\n",
       "      <td>40,100</td>\n",
       "      <td>140.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>150,0,1888</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2235792-A/G</th>\n",
       "      <td>1</td>\n",
       "      <td>2235792</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>436.66</td>\n",
       "      <td>QDfilterSNV</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>93,17</td>\n",
       "      <td>110</td>\n",
       "      <td>99</td>\n",
       "      <td>145,0,1894</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0/0</td>\n",
       "      <td>50,50</td>\n",
       "      <td>100.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>100,0,100</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2235901-A/G</th>\n",
       "      <td>1</td>\n",
       "      <td>2235901</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.89</td>\n",
       "      <td>LowQual</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,1</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>3,0,197</td>\n",
       "      <td>0.09</td>\n",
       "      <td>1/1</td>\n",
       "      <td>200,200</td>\n",
       "      <td>400.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>100,0,100</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2239999-A/G,T</th>\n",
       "      <td>1</td>\n",
       "      <td>2239999</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G,T</td>\n",
       "      <td>24.89</td>\n",
       "      <td>LowQual</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,1,4</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>3,0,197</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0/1</td>\n",
       "      <td>100,20,100</td>\n",
       "      <td>220.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1000,500,1000</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2235501-A/GGT</th>\n",
       "      <td>1</td>\n",
       "      <td>2235501</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>GGT</td>\n",
       "      <td>24.89</td>\n",
       "      <td>LowQual</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,1</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>3,0,197</td>\n",
       "      <td>0.09</td>\n",
       "      <td>1/1</td>\n",
       "      <td>200,200</td>\n",
       "      <td>400.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>100,0,100</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2239901-GA/G</th>\n",
       "      <td>1</td>\n",
       "      <td>2239901</td>\n",
       "      <td>.</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>24.89</td>\n",
       "      <td>LowQual</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,1</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>3,0,197</td>\n",
       "      <td>0.09</td>\n",
       "      <td>1/1</td>\n",
       "      <td>200,200</td>\n",
       "      <td>400.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>100,0,100</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                CHROM      POS ID REF  ALT     QUAL       FILTER  \\\n",
       "                                                                   \n",
       "1:1000-A/G       1     1000     .  A   G    24.89    LowQual       \n",
       "1:2234385-C/T    1     2234385  .  C   T    9953.67  PASS          \n",
       "1:2235243-C/T    1     2235243  .  C   T    82.06    PASS          \n",
       "1:2235792-A/G    1     2235792  .  A   G    436.66   QDfilterSNV   \n",
       "1:2235901-A/G    1     2235901  .  A   G    24.89    LowQual       \n",
       "1:2239999-A/G,T  1     2239999  .  A   G,T  24.89    LowQual       \n",
       "1:2235501-A/GGT  1     2235501  .  A   GGT  24.89    LowQual       \n",
       "1:2239901-GA/G   1     2239901  .  GA  G    24.89    LowQual       \n",
       "\n",
       "                         FORMAT INFO      26PL1207                   \\\n",
       "                                  AC LOLZ       GT      AD   DP  GQ   \n",
       "1:1000-A/G       GT:AD:DP:GQ:PL  2    9    0/1      10,1    11   3    \n",
       "1:2234385-C/T    GT:AD:DP:GQ:PL  1    90   0/1      5,11    16   99   \n",
       "1:2235243-C/T    GT:AD:DP:GQ:PL  0    0    0/1      4,1     5    16   \n",
       "1:2235792-A/G    GT:AD:DP:GQ:PL  0    0    0/1      93,17   110  99   \n",
       "1:2235901-A/G    GT:AD:DP:GQ:PL  0    0    0/1      10,1    11   3    \n",
       "1:2239999-A/G,T  GT:AD:DP:GQ:PL  0    0    0/1      10,1,4  15   3    \n",
       "1:2235501-A/GGT  GT:AD:DP:GQ:PL  0    0    0/1      10,1    11   3    \n",
       "1:2239901-GA/G   GT:AD:DP:GQ:PL  0    0    0/1      10,1    11   3    \n",
       "\n",
       "                                  22MI1099                           \\\n",
       "                         PL    AB       GT          AD     DP    GQ   \n",
       "1:1000-A/G       3,0,197     0.09  ./.      NaN        NaN    NaN     \n",
       "1:2234385-C/T    289,0,103   0.69  0/1      10,1        11.0   3.0    \n",
       "1:2235243-C/T    16,0,102    0.20  0/1      40,100      140.0  50.0   \n",
       "1:2235792-A/G    145,0,1894  0.15  0/0      50,50       100.0  89.0   \n",
       "1:2235901-A/G    3,0,197     0.09  1/1      200,200     400.0  90.0   \n",
       "1:2239999-A/G,T  3,0,197     0.07  0/1      100,20,100  220.0  90.0   \n",
       "1:2235501-A/GGT  3,0,197     0.09  1/1      200,200     400.0  90.0   \n",
       "1:2239901-GA/G   3,0,197     0.09  1/1      200,200     400.0  90.0   \n",
       "\n",
       "                                      \n",
       "                            PL    AB  \n",
       "1:1000-A/G       NaN           NaN    \n",
       "1:2234385-C/T    3,0,197        0.09  \n",
       "1:2235243-C/T    150,0,1888     0.71  \n",
       "1:2235792-A/G    100,0,100      0.50  \n",
       "1:2235901-A/G    100,0,100      0.50  \n",
       "1:2239999-A/G,T  1000,500,1000  0.09  \n",
       "1:2235501-A/GGT  100,0,100      0.50  \n",
       "1:2239901-GA/G   100,0,100      0.50  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "    \n",
    "\n",
    "vcf = vcf2dataframe(t, genotype_level=True, \n",
    "              info_level=True, UID=True)\n",
    "vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "class VCF(object):\n",
    "    ''' A VCF file stored as a Pandas DataFrame\n",
    "    \n",
    "    Atrributes:\n",
    "        vcf: vcf file to be converted to a Pandas DataFrame or a VCF object\n",
    "        convert: if True convert to pandas DataFrame, False assumes vcf is already a VCF object\n",
    "        genotype_level: place the genotype information into a second level column index\n",
    "        info_level: place the info IDs into a second level column index\n",
    "        UID: rename index to a unique variant identifier\n",
    "    '''\n",
    "    def __init__(self, vcf, convert=True, genotype_level=True, info_level=True, UID=True):\n",
    "        \n",
    "        if convert:\n",
    "            self.vcf = vcf2dataframe(vcf, genotype_level=genotype_level,\n",
    "                                     info_level=info_level, UID=UID)\n",
    "        else:\n",
    "            self.vcf = vcf\n",
    "        \n",
    "        \n",
    "    def get_genotype(self, gen):\n",
    "        ''' Access specific genotype information across samples\n",
    "            in the vcf.\n",
    "        Args:\n",
    "            gen: genotype attribute of interest in string format e.g 'DP'\n",
    "        '''\n",
    "        num_info = self.vcf['INFO'].columns.shape[0]\n",
    "        genotype = self.vcf.ix[:,9+num_info:]\n",
    "        return genotype.xs(gen, level=1, axis=1)\n",
    "\n",
    "    \n",
    "    def get_info(self, info):\n",
    "        ''' Return INFO field of interest e.g. 'AC'\n",
    "        '''\n",
    "        return self.vcf['INFO'][info]\n",
    "\n",
    "\n",
    "\n",
    "class FilterVCF(VCF):\n",
    "    ''' A VCF file which can be readily filtered\n",
    "    \n",
    "    Attributes:\n",
    "        vcf: Pandas DataFrame VCF \n",
    "    '''\n",
    "    def __init__(self, vcf):\n",
    "        #self.vcf = vcf\n",
    "        VCF.__init__(self, vcf, convert=False)\n",
    "\n",
    "\n",
    "    def subset(self, sams, exclude_ref=False, remove_uncalled=True):\n",
    "        ''' Subset a multisample VCF by a given samples.\n",
    "        Args:\n",
    "            vcf: Pandas DataFrame VCF\n",
    "            sams: list of samples to subset the vcf for\n",
    "            exlude_ref: remove variant if all GT values for subset are 0/0\n",
    "            remove_uncalled: remove variant if all GT values for subset are ./.\n",
    "\n",
    "        Returns:\n",
    "            subsetted Pandas DataFrame VCF\n",
    "        '''\n",
    "        # split variant and genotype information \n",
    "        genotype = self.vcf[sams]\n",
    "        num_info = self.vcf['INFO'].columns.shape[0]\n",
    "        variant = self.vcf.ix[:,:8+num_info]\n",
    "\n",
    "        GT = genotype.xs('GT', level=1, axis=1)\n",
    "        uncalled= []\n",
    "\n",
    "        if remove_uncalled:\n",
    "            uncalled = GT[GT[sams] == './.'].dropna().index.tolist() \n",
    "\n",
    "        if exclude_ref:\n",
    "            uncalled += GT[GT[sams] == '0/0'].dropna().index.tolist() \n",
    "\n",
    "        sub = pd.concat([variant, genotype], axis=1)\n",
    "        self.vcf = sub.drop(uncalled)\n",
    "        return self.vcf\n",
    "\n",
    "    \n",
    "    def filter_genotype(self, minDP=None, minGQ=None, minAB=None):\n",
    "        ''' Filter for variants in which all the samples in the given vcf \n",
    "            meet the minimum genotype values given.\n",
    "        \n",
    "        Args:\n",
    "            minDP: minimum variant depth\n",
    "            minGQ: minimum genotype quality\n",
    "            minAB: minimum allele balance\n",
    "        \n",
    "        Notes:\n",
    "            Doesn't handle multiallelic information properly and\n",
    "            will filter for this first ALT value e.g. if DP = 12,1,100\n",
    "            it will be filtered out even if minDP=30.\n",
    "        '''\n",
    "        # split variant and genotype information\n",
    "        num_info = self.vcf['INFO'].columns.shape[0]\n",
    "        variant = self.vcf.ix[:,:8+num_info]\n",
    "        genotype = self.vcf.ix[:,9+num_info:]\n",
    "\n",
    "        # store all variants that don't meet the minimum value given for the args here\n",
    "        below_min = []\n",
    "        \n",
    "        if minDP:\n",
    "            DP = genotype.xs('DP', level=1, axis=1).fillna(0)\n",
    "            above_min = DP[DP >= minDP] \n",
    "            below_min += DP[above_min.isnull().any(axis=1)].index.tolist()\n",
    "            \n",
    "        if minGQ:\n",
    "            GQ = genotype.xs('GQ', level=1, axis=1).fillna(0)\n",
    "            above_min = GQ[GQ >= minGQ]\n",
    "            below_min +=  GQ[above_min.isnull().any(axis=1)].index.tolist()\n",
    "        \n",
    "        if minAB:\n",
    "            AB = genotype.xs('AB', level=1, axis=1).fillna(0)\n",
    "            above_min = AB[AB >= minAB]\n",
    "            below_min +=  AB[above_min.isnull().any(axis=1)].index.tolist()\n",
    "        \n",
    "        # remove variants that don't meet the requirements from the vcf\n",
    "        self.vcf = self.vcf.drop(below_min)\n",
    "        return self.vcf\n",
    "    \n",
    "    \n",
    "    def filter_info(self, field, value):\n",
    "        ''' Filter for variants that are above the given value\n",
    "            (if value is number) or are equal to the given value\n",
    "            (if value is string).\n",
    "            \n",
    "        Args:\n",
    "            field: INFO field of interest\n",
    "            value: string or int value to test the field with\n",
    "            \n",
    "        Notes:\n",
    "            Doesn't handle multiallelic information properly and\n",
    "            will filter any thing that has this e.g. if AC = 12,34\n",
    "            it will be filtered out even if value=1 as the AC with\n",
    "            a comma cant be converted to an int so becomes np.nan\n",
    "        '''\n",
    "        if isinstance(value, int) or isinstance(value, float):\n",
    "            mask = pd.to_numeric(self.vcf['INFO'][field], errors='coerrce') >= value\n",
    "            self.vcf = self.vcf[mask]\n",
    "            return self.vcf\n",
    "        \n",
    "        elif isinstance(value, str):\n",
    "            mask= self.vcf['INFO'][field] == value\n",
    "            self.vcf = self.vcf[mask]\n",
    "            return self.vcf\n",
    "        \n",
    "    \n",
    "    def remove_indels(self):\n",
    "        ''' Remove indels from vcf.\n",
    "        '''\n",
    "        alt_mask = (self.vcf.ALT.str.len() == 1) | (self.vcf.ALT.str.contains(','))\n",
    "        ref_mask = (self.vcf.REF.str.len() == 1) | (self.vcf.REF.str.contains(','))\n",
    "        return(self.vcf[alt_mask & ref_mask])\n",
    "    \n",
    "    \n",
    "    def biallelic(self):\n",
    "        ''' Filter for biallelic variants only.\n",
    "        '''\n",
    "        self.vcf = self.vcf[self.vcf.ALT.str.split(',').str.len() == 1]\n",
    "        return self.vcf\n",
    "    \n",
    "    \n",
    "    def multiallelic(self):\n",
    "        ''' Filter for multiallelic variants only.\n",
    "        '''\n",
    "        self.vcf = self.vcf[self.vcf.ALT.str.split(',').str.len() > 1]\n",
    "        return self.vcf\n",
    "    \n",
    "    \n",
    "    def positions(self, positions, include=True):\n",
    "        ''' Include or exclude variants in the given position(s)\n",
    "            or position ranges.\n",
    "        \n",
    "        Args:\n",
    "            positions: a position, position ranges or list of the two e.g.\n",
    "                            ['1:2234385', '1:2235901-2240000']\n",
    "            include: include in the vcf if true, otherwise exclude\n",
    "        '''\n",
    "        positions = positions if isinstance(positions, list) else [positions]\n",
    "        \n",
    "        # get the variants named in positions\n",
    "        selected_variants = []\n",
    "        \n",
    "        for pos in positions:\n",
    "            \n",
    "            # place a dash in pos if not present\n",
    "            if \"-\" not in pos:\n",
    "                split = pos.split(\":\")\n",
    "                pos = \"{}:{}-{}\".format(split[0], split[1], split[1])\n",
    "\n",
    "            chrom, start, end = [int(x) for x in re.split(r'[:-]', pos)]\n",
    "            mask = (self.vcf['CHROM'] == chrom) & (self.vcf['POS'] >= start) & (self.vcf['POS'] <= end)\n",
    "            variants = self.vcf[mask].index.tolist()\n",
    "            \n",
    "            selected_variants.append(variants)\n",
    "            \n",
    "        # flatten\n",
    "        selected_variants = [y for x in selected_variants for y in x]\n",
    "        \n",
    "        \n",
    "        if not include:\n",
    "            selected_variants = list(set(self.vcf.index.tolist()) - set(selected_variants))\n",
    "            \n",
    "        self.vcf = self.vcf.loc[FilterVCF.natural_sort(selected_variants)] \n",
    "\n",
    "        return self.vcf\n",
    "    \n",
    "    \n",
    "    \n",
    "    @staticmethod\n",
    "    def natural_sort(l): \n",
    "        convert = lambda text: int(text) if text.isdigit() else text.lower() \n",
    "        alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ] \n",
    "        return sorted(l, key = alphanum_key)\n",
    "    \n",
    "    @staticmethod\n",
    "    def pos2range(pos): \n",
    "        if \"-\" not in pos:\n",
    "            split = pos.split(\":\")\n",
    "            pos = \"{}:{}-{}\".format(split[0], split[1], split[1])\n",
    "            return pos\n",
    "        else:\n",
    "            pos\n",
    "        \n",
    "        \n",
    "    \n",
    "    \n",
    "        \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AC', 'AF', 'AN', 'BaseQRankSum', 'DB', 'DP', 'DS', 'Dels', 'ExcessHet', 'FS', 'HaplotypeScore', 'InbreedingCoeff', 'MLEAC', 'MLEAF', 'MQ', 'MQ0', 'MQRankSum', 'QD', 'RPA', 'RU', 'ReadPosRankSum', 'SOR', 'STR', 'set']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:220: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:253: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "expected string or bytes-like object",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-257-8d2b7b725ae0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m \u001b[0mtest_vcf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpositions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrange_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minclude\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-254-990459d73aa5>\u001b[0m in \u001b[0;36mpositions\u001b[0;34m(self, positions, include)\u001b[0m\n\u001b[1;32m    185\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m             \u001b[0mpos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFilterVCF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpos2range\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m             \u001b[0mchrom\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mre\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mr'[:-]'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m             \u001b[0mmask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvcf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'CHROM'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mchrom\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvcf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvcf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m             \u001b[0mvariants\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvcf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.5/re.py\u001b[0m in \u001b[0;36msplit\u001b[0;34m(pattern, string, maxsplit, flags)\u001b[0m\n\u001b[1;32m    201\u001b[0m     \u001b[0;32mand\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mremainder\u001b[0m \u001b[0mof\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mstring\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mreturned\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mfinal\u001b[0m \u001b[0melement\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    202\u001b[0m     of the list.\"\"\"\n\u001b[0;32m--> 203\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_compile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstring\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmaxsplit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    204\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mfindall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstring\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: expected string or bytes-like object"
     ]
    }
   ],
   "source": [
    "v = VCF(o, genotype_level=True, \n",
    "    info_level=True, UID=True)\n",
    "\n",
    "\n",
    "test_vcf = FilterVCF(v.vcf)\n",
    "\n",
    "#mask = (test_vcf.vcf.REF.str.len() == 1) & (test_vcf.vcf.ALT.str.len() == 1)\n",
    "#test_vcf.vcf[mask]\n",
    "#test_vcf.vcf[test_vcf.vcf.ALT.str.split(',').str.len() == 1]\n",
    "#test_vcf.remove_indels()\n",
    "#test_vcf.vcf['INFO'].apply(pd.to_numeric)\n",
    "#test_vcf.get_genotype('DP')['22MI1099']\n",
    "\n",
    "range_pos = ['20:45354737-45364737', '1:2160444']\n",
    "\n",
    "\n",
    "test_vcf.positions(range_pos, include=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AC', 'LOLZ']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:220: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:253: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th colspan=\"2\" halign=\"left\">INFO</th>\n",
       "      <th colspan=\"6\" halign=\"left\">22MI1099</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AC</th>\n",
       "      <th>LOLZ</th>\n",
       "      <th>GT</th>\n",
       "      <th>AD</th>\n",
       "      <th>DP</th>\n",
       "      <th>GQ</th>\n",
       "      <th>PL</th>\n",
       "      <th>AB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1:2234385-C/T</th>\n",
       "      <td>1</td>\n",
       "      <td>2234385</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9953.67</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,1</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3,0,197</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:2235901-A/G</th>\n",
       "      <td>1</td>\n",
       "      <td>2235901</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>24.89</td>\n",
       "      <td>LowQual</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>200,200</td>\n",
       "      <td>400.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>100,0,100</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              CHROM      POS ID REF ALT     QUAL   FILTER          FORMAT  \\\n",
       "                                                                            \n",
       "1:2234385-C/T  1     2234385  .  C   T   9953.67  PASS     GT:AD:DP:GQ:PL   \n",
       "1:2235901-A/G  1     2235901  .  A   G   24.89    LowQual  GT:AD:DP:GQ:PL   \n",
       "\n",
       "              INFO      22MI1099                                         \n",
       "                AC LOLZ       GT       AD     DP    GQ         PL    AB  \n",
       "1:2234385-C/T  1    90   0/1      10,1     11.0   3.0   3,0,197    0.09  \n",
       "1:2235901-A/G  0    0    1/1      200,200  400.0  90.0  100,0,100  0.50  "
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "v = VCF(t, genotype_level=True, \n",
    "    info_level=True, UID=True)\n",
    "\n",
    "\n",
    "test_vcf = FilterVCF(v.vcf)\n",
    "test_vcf.get_genotype('DP')\n",
    "\n",
    "#test_vcf.filter_genotype(minDP=10, minAB=0.1)#, minAB=0.1)\n",
    "test_vcf.vcf#.loc[['1:1000-A/G']]\n",
    "\n",
    "test_vcf.get_genotype('DP')\n",
    "\n",
    "test_vcf.subset(['22MI1099'], remove_uncalled=True, exclude_ref=True)\n",
    "#test_vcf.genotype(minDP=12, minGQ=51)\n",
    "test_vcf.vcf.loc[['1:2234385-C/T', '1:2235901-A/G']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1:2235243-C/T', '1:2235501-A/GGT', '1:2239901-GA/G', '1:2239999-A/G,T'}"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "set(test_vcf.vcf.index.tolist()) - set(['1:2234385-C/T', '1:2235901-A/G'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
